content/ko/docs/tasks/run-application/run-replicated-stateful-application.md

```diff
@@ -0,0 +1,520 @@
+접근이 가능하다.
+
+`mysql-read`라고 불리우는 클라이언트 서비스는 고유의 클러스터IP를 가지며, Ready 상태인 모든
+MySQL 파드들에게 커넥션들을 분배해주는 일반적인 서비스이다. 잠재적인 엔드포인트들의 집합은
+주 MySQL 서버와 해당 레플리카들을 포함한다.
+
+오직 읽기 쿼리들만 로드-밸런싱된 클라이언트 서비스를 이용할 수 있다는 사실에 주목하자.
+하나의 주 MySQL 서버만이 존재하기 떄문에, 클라인언트들은 쓰기 작업을 실행하기 위해서 
+주 MySQL 파드에 (헤드리스 서비스 안에 존재하는 DNS 엔드리를 통해) 직접적으로 접근해야 한다.
+
+### 스테이트풀셋 생성하기 {#statefulset}
+
+마지막으로, 다음 YAML 설정 파일로부터 스테이트풀셋을 생성한다:
+
+{{< codenew file="application/mysql/mysql-statefulset.yaml" >}}
+
+```shell
+kubectl apply -f https://k8s.io/examples/application/mysql/mysql-statefulset.yaml
+```
+
+다음을 실행하여, 초기화되는 프로세스들을 확인할 수 있다:
+
+```shell
+kubectl get pods -l app=mysql --watch
+```
+
+잠시 뒤에, 3개의 파드들이 `Running` 상태가 되는 것을 볼 수 있다:
+
+```
+NAME      READY     STATUS    RESTARTS   AGE
+mysql-0   2/2       Running   0          2m
+mysql-1   2/2       Running   0          1m
+mysql-2   2/2       Running   0          1m
+```
+
+**Ctrl+C**를 입력하여 watch를 종료하자.
+
+{{< note >}}
+만약 아무 진행상황도 보이지 않는다면, [시작하기 전에](#before-you-begin) 에 언급된
+동적 퍼시스턴트볼륨 프로비저너가 활성화되어 있는지 확인한다.
+{{< /note >}}
+
+해당 메니페스트는 스테이트풀셋의 일부분인 스테이트풀 파드들을 관리하기 위해 다양한 기법들이
+적용되었다. 다음 섹션에서는 스테트풀셋이 파드들을 생성하면서 일어나는 일부 기법을 설명한다.
+
+## 스테이트풀 파드 초기화 이해하기
+
+스테이트풀셋 컨트롤러는 파드들의 인덱스에 따라 하나씩 시작시킨다.
+컨트롤러는 다음 파드 생성 이전에 각 파드가 Ready 상태가 되었다고 알려줄 떄까지 기다린다.
+
+추가적으로, 컨트롤러는 각 파드들에게 `<스테이트풀셋 이름>-<인덱스>`의 고유하고 안정적인 이름을
+부여하는데, 결과적으로 파드들은 `mysql-0`, `mysql-1`, 그리고 `mysql-2` 라는 이름을 가지게 된다.
+
+스테이트풀셋 매니페스트의 파드 템플릿은 해당 속성들을 통해 순차적인 MySQL 복제의 시작을 수행한다.
+
+### 설정 생성하기
+
+파드 스펙의 컨테이너를 시작하기 전에, 파드는 순서가 정의되어 있는
+[초기화 컨테이너](/ko/docs/concepts/workloads/pods/init-containers/)들을 먼저 실행시긴다.
+
+`init-mysql`라는 이름의 첫번째 초기화 컨테이너는, 인덱스에 따라 특별한 MySQL 설정 파일을 생성한다.
+
+스크립트는 인덱스를 `hostname` 명령으로 반환되는 파드 이름의 마지막 부분에서 추출하여 결정한다.
+그리고 인덱스(이미 사용된 값들을 피하기 위한 오프셋 숫자와 함께)를 MySQL의 `conf.d` 디렉토리의
+`server-id.cnf` 파일에 저장한다.
+이는 스테이트풀셋에게서 제공된 고유하고, 안정적인 신원을 같은 속성을
+필요로 하는 MySQL 서버 ID의 형태로 바꾸어준다.
+
+또한 `init-mysql` 컨테이너의 스크립트는 컨피그맵을 `conf.d`로 복사하여, `primary.cnf`이나
+`replica.cnf`을 적용한다.
+이 예제의 토폴로지가 하나의 주 MySQL 서버와 일정 수의 레플리카들로 이루어져 있기 때문에,
+스크립트는 `0` 인덱스를 주 서버로, 그리고 나머지 값들은 레플리카로 지정한다.
+스테이트풀셋 컨트롤러의
+[디플로이먼트와 스케일링 보증](/ko/docs/concepts/workloads/controllers/statefulset/#deployment-and-scaling-guarantees)과
+합쳐지면, 복제를 위한 레플리카들을 생성하기 전에 주 MySQL 서버가 Ready 상태가 되도록 보장할 수 있다.
+
+### 기존 데이터 복제
+
+일반적으로는, 레플리카에 새로운 파드가 들어오면, 주 MySQL 서버가 이미 데이터를 가지고 있다고 가정해야 한다.
+또한 복제 로그가 첫 시작점부터의 로그들을 다 가지고 있지는 않을 수 있다고 가정해야 한다.
+이러한 보수적인 가정들은 스테이트풀셋이 초기 크기로 고정되어 있는 것보다, 시간에 따라
+확장/축소하게 할 수 있도록 하는 중요한 열쇠가 된다.
+
+`clone-mysql`라는 이름의 두번째 초기화 컨테이너는, 퍼시스턴트볼륨에서 처음 초기화된
+레플리카 파드에 복제 작업을 수행한다.
+이 말은 다른 실행중인 파드로부터 모든 데이터들을 복제하기 때문에,
+로컬의 상태가 충분히 일관성을 유지하고 있어서 주 서버에서부터 복제를 시작할 수 있다는 의미이다.
+
+MySQL 자체가 이러한 메커니즘을 제공해주지는 않기 떄문에, 이 예제에서는 XtraBackup이라는
+유명한 오픈소스 툴을 사용한다.
+복제 중에는, 복제 대상의 MySQL 서버가 감소된 퍼포먼스로 부담이 가중될 수 있다.
+주 MySQL 서버의 이러한 충격을 최소화하기 위해, 스크립트는 각 파드가 자신의 인덱스보다 
+하나 작은 파드로부터 복제하도록 지시한다.
+이것은 스테이트풀셋 컨트롤러가 파드 `N+1`을 실행하기 전에 항상 파드 `N`이 실행중이라는 것을
+보장해주기 때문이다.

```

Comment:
```suggestion
이것이 정상적으로 동작하는 이유는 스테이트풀셋 컨트롤러가 파드 `N+1`을 실행하기 전에 항상 파드 `N`이 Ready 상태라는 것을
보장하기 때문이다.
```

원문을 고려하여 수정해보았습니다.
```
This works because the StatefulSet controller always ensures Pod `N` is
Ready before starting Pod `N+1`.
```